juniper/src/integrations/chrono.rs

```diff
@@ -78,54 +146,82 @@ where
+
+#[graphql_scalar(
+    description = "LocalDateTime",
+    description = "Combined date and time (without time zone) in `yyyy-MM-dd \
+                   HH:mm:ss` format.\
+                   \n\n\
+                   See also [`chrono::NaiveDateTime`][1] for details.\
+                   \n\n\
+                   [1]: https://docs.rs/chrono/*/chrono/naive/struct.NaiveDateTime.html"
+)]
+impl<S> GraphQLScalar for NaiveDateTime
 where
     S: ScalarValue,
 {
     fn resolve(&self) -> Value {
-        Value::scalar(self.format("%Y-%m-%d").to_string())
+        Value::scalar(self.format(LOCAL_DATE_TIME_FORMAT).to_string())
     }
 
-    fn from_input_value(v: &InputValue) -> Result<NaiveDate, String> {
+    fn from_input_value(v: &InputValue) -> Result<NaiveDateTime, String> {
         v.as_string_value()
             .ok_or_else(|| format!("Expected `String`, found: {}", v))
             .and_then(|s| {
-                NaiveDate::parse_from_str(s, "%Y-%m-%d")
-                    .map_err(|e| format!("Failed to parse `NaiveDate`: {}", e))
+                Self::parse_from_str(s, LOCAL_DATE_TIME_FORMAT)
+                    .map_err(|e| format!("Invalid `LocalDateTime`: {}", e))
             })
     }
 
     fn from_str<'a>(value: ScalarToken<'a>) -> ParseScalarResult<'a, S> {
-        if let ScalarToken::String(value) = value {
-            Ok(S::from(value.to_owned()))
+        if let ScalarToken::String(s) = value {
+            Ok(S::from(s.to_owned()))
         } else {
             Err(ParseError::UnexpectedToken(Token::Scalar(value)))
         }
     }
 }
 
-#[cfg(feature = "scalar-naivetime")]
-#[crate::graphql_scalar(description = "NaiveTime")]
-impl<S> GraphQLScalar for NaiveTime
+/// Format of a [`DateTime` scalar][1].
+///
+/// [1]: https://graphql-scalars.dev/docs/scalars/date-time
+const DATE_TIME_FORMAT: &str = "%FT%TZ";
+
+#[graphql_scalar(
+    name = "DateTime",
+    description = "Combined date and time (with time zone) in [RFC 3339][0] \
+                   format.\
+                   \n\n\
+                   Represents a description of an exact instant on the \
+                   time-line (such as the instant that a user account was \
+                   created).\
+                   \n\n\
+                   [`DateTime` scalar][1] compliant.\
+                   \n\n\
+                   See also [`chrono::DateTime`][2]`<`[`FixedOffset`][3]`>` \
+                   for details.\
+                   \n\n\
+                   [0]: https://datatracker.ietf.org/doc/html/rfc3339#section-5.6\n\
+                   [1]: https://graphql-scalars.dev/docs/scalars/date-time\n\
+                   [2]: https://docs.rs/chrono/*/chrono/struct.DateTime.html\n\",
+                   [3]: https://docs.rs/chrono/*/chrono/offset/struct.FixedOffset.html",
+    specified_by_url = "https://graphql-scalars.dev/docs/scalars/date-time"
+)]
+impl<S> GraphQLScalar for DateTime<FixedOffset>
```

Comment:
What we do really want here is
```rust
impl<S: ScalarValue, Tz: TimeZone> GraphQLScalar for DateTime<Tz>
```
but `#[graphql_scalar]` macro doesn't allows us to make generic impls.

So, let's first rewrite `#[graphql_scalar]` so we can do that.

Another problem arosen here is that `chrono` is very bad at relations between `TimeZone`s. We have no relations like (`From`) between `Utc`, `FixedOffset` and `Local`, and the only thing how we can create `Tz: TimeZone` is `TimeZone::from_offset()`, where `Offset` is another generic parameter.

So we should write up some `FromUtcOffset` (or similar) machinery here, to wire them up, so, after parsing `DateTime<FixedOffset>` we can convert it into arbitrary `DateTime<Tz>`.

Reply:
@tyranron 

> So we should write up some `FromUtcOffset` (or similar) machinery here, to wire them up, so, after parsing `DateTime<FixedOffset>` we can convert it into arbitrary `DateTime<Tz>`.

All in all the main problem is that to convert into arbitrary `DateTime<Tz>` you need an instance of a type, not just type itself. So even with custom machinery we can support only `FixedOffset` and `Utc`. [`chrono-tz`][1] on the other hand, uses big enum instead of separate types to represent all timezones, so our users will have to implement needed timezones manually.

I propose different approach: we allow only `DateTime<Utc>` and `DateTime<FixedOffsets>` as inputs, while any `DateTime<Tz>` as output. So usage will look something like that:

```rust
#[graphql_object]
impl QueryRoot {
    fn transform(input: DateTime<Utc>) -> DateTime<chrono_tz::Tz> {
        input.with_timezone(&chrono_tz::CET)
    }
}
```

The only downside is that we won't be able to use macros to implement scalar and fallback to manual implementation. This happens because with don't have a mechanism to enforce different trait bounds on different resolvers.




[1]: https://docs.rs/chrono-tz/0.6.1/chrono_tz/index.html